File: app/src/main/java/com/aar/app/wsp/data/room/GameDatabase.java
package com.aar.app.wsp.data.room;

import androidx.sqlite.db.SupportSQLiteDatabase;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import android.content.Context;
import androidx.annotation.NonNull;

import com.aar.app.wsp.data.xml.WordThemeDataXmlLoader;
import com.aar.app.wsp.model.GameTheme;
import com.aar.app.wsp.model.UsedWord;
import com.aar.app.wsp.model.Word;

import java.util.concurrent.Executors;

@Database(entities = {Word.class, GameTheme.class, UsedWord.class}, version = 1)
public abstract class GameDatabase extends RoomDatabase {
    private static final String DB_NAME = "game_data.db";

    private static GameDatabase INSTANCE = null;

    public static GameDatabase getInstance(@NonNull final Context context) {
        if (INSTANCE == null) {
            INSTANCE = Room.databaseBuilder(context, GameDatabase.class, DB_NAME)
                    .addCallback(new Callback() {
                        @Override
                        public void onCreate(@NonNull SupportSQLiteDatabase db) {
                            Executors.newSingleThreadScheduledExecutor()
                                    .execute(() -> prepopulateDatabase(context));
                        }
                    })
                    .build();
        }
        return INSTANCE;
    }

    private static void prepopulateDatabase(@NonNull Context context) {
        WordThemeDataXmlLoader dataXmlLoader = new WordThemeDataXmlLoader(context);

        GameDatabase gameDb = GameDatabase.getInstance(context);
        gameDb.getWordDataSource().insertAll(dataXmlLoader.getWords());
        gameDb.getGameThemeDataSource().insertAll(dataXmlLoader.getGameThemes());

        dataXmlLoader.release();
    }

    public abstract WordDataSource getWordDataSource();
    public abstract UsedWordDataSource getUsedWordDataSource();
    public abstract GameThemeDataSource getGameThemeDataSource();
}

File: app/src/main/java/com/aar/app/wsp/data/room/GameDatabase.kt
package com.aar.app.wsp.data.room

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.sqlite.db.SupportSQLiteDatabase
import com.aar.app.wsp.data.xml.WordThemeDataXmlLoader
import com.aar.app.wsp.model.GameTheme
import com.aar.app.wsp.model.UsedWord
import com.aar.app.wsp.model.Word
import java.util.concurrent.Executors

@Database(entities = [Word::class, GameTheme::class, UsedWord::class], version = 1)
abstract class GameDatabase : RoomDatabase() {
    abstract val wordDataSource: WordDataSource
    abstract val usedWordDataSource: UsedWordDataSource
    abstract val gameThemeDataSource: GameThemeDataSource

    companion object {
        private const val DB_NAME = "game_data.db"
        private var INSTANCE: GameDatabase? = null
        fun getInstance(context: Context): GameDatabase {
            if (INSTANCE == null) {
                INSTANCE = Room.databaseBuilder(context, GameDatabase::class.java, DB_NAME)
                    .addCallback(object : Callback() {
                        override fun onCreate(db: SupportSQLiteDatabase) {
                            Executors.newSingleThreadScheduledExecutor()
                                .execute { prepopulateDatabase(context) }
                        }
                    })
                    .build()
            }
            return INSTANCE!!
        }

        private fun prepopulateDatabase(context: Context) {
            val dataXmlLoader = WordThemeDataXmlLoader(context)
            val gameDb = getInstance(context)
            gameDb.wordDataSource.insertAll(dataXmlLoader.words)
            gameDb.gameThemeDataSource.insertAll(dataXmlLoader.gameThemes)
            dataXmlLoader.release()
        }
    }
}